be/src/exec/workgroup/work_group.h

```diff
@@ -10,82 +10,132 @@
     virtual WorkGroupPtr pick_next() = 0;
-};
 
-class CpuWorkGroupQueue final : public WorkGroupQueue {
-public:
-    CpuWorkGroupQueue() = default;
-    ~CpuWorkGroupQueue() = default;
-    void add(const WorkGroupPtr& wg) override {}
-    void remove(const WorkGroupPtr& wg) override {}
-    WorkGroupPtr pick_next() override { return nullptr; }
+    virtual void close() = 0;
 };
 
 class IoWorkGroupQueue final : public WorkGroupQueue {
 public:
     IoWorkGroupQueue() = default;
-    ~IoWorkGroupQueue() = default;
+    ~IoWorkGroupQueue() override = default;
+
     void add(const WorkGroupPtr& wg) override {}
     void remove(const WorkGroupPtr& wg) override {}
-    WorkGroupPtr pick_next() override { return nullptr; }
+    WorkGroupPtr pick_next() override { return nullptr; };
+
+    StatusOr<PriorityThreadPool::Task> pick_next_task();
+    bool try_offer_io_task(WorkGroupPtr wg, const PriorityThreadPool::Task& task);
+    void close() override;
 };
 
 class WorkGroupManager;
-
-enum WorkGroupType {
-    WG_NORMAL = 0,   // normal work group, maybe added to the BE dynamically
-    WG_DEFAULT = 1,  // default work group
-    WG_REALTIME = 2, // realtime work group, maybe reserved beforehand
-};
+using WorkGroupType = TWorkGroupType::type;
 // WorkGroup is the unit of resource isolation, it has {CPU, Memory, Concurrency} quotas which limit the
 // resource usage of the queries belonging to the WorkGroup. Each user has be bound to a WorkGroup, when
 // the user issues a query, then the corresponding WorkGroup is chosen to manage the query.
 class WorkGroup {
 public:
-    WorkGroup(const std::string& name, int id, size_t cpu_limit, size_t memory_limit, size_t concurrency,
-              WorkGroupType type);
+    WorkGroup(const std::string& name, int64_t id, int64_t version, size_t cpu_limit, double memory_limit,
+              size_t concurrency, WorkGroupType type);
+    WorkGroup(const TWorkGroup& twg);
     ~WorkGroup() = default;
 
+    TWorkGroup to_thrift() const;
+    TWorkGroup to_thrift_verbose() const;
     void init();
 
-    starrocks::MemTracker* mem_tracker() { return _mem_tracker.get(); }
-    starrocks::pipeline::DriverQueue* driver_queue() { return _driver_queue.get(); }
+    MemTracker* mem_tracker() { return _mem_tracker.get(); }
+    pipeline::DriverQueue* driver_queue() { return _driver_queue.get(); }
+
+    int64_t id() const { return _id; }
+
+    int64_t version() const { return _version; }
+
+    const std::string& name() const { return _name; }
 
-    int id() const { return _id; }
-    int get_cpu_priority() {
-        // TODO: implement cpu priority computation
-        return 0;
+    static constexpr int64 DEFAULT_WG_ID = 0;
+    static constexpr int64 DEFAULT_VERSION = 0;
+    bool try_offer_io_task(const PriorityThreadPool::Task& task);
+
+public:
+    // increase num_driver when the driver is attached to the workgroup
+    void increase_num_drivers() {
+        ++_num_drivers;
+        ++_acc_num_drivers;
```

Comment:
what is the difference between `_num_drivers` and `_acc_num_drivers`?

Reply:
1. _num_drivers is the number of active drivers, so when a driver attach to the workgroup, it increases by 1, when a driver detach from the driver, it decrease by 1.
2.  _acc_num_drivers is the number of drivers attached to this workgroup since be's startup.